src/main/java/graphql/annotations/ExtensionDataFetcherWrapper.java

```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright 2016 Yurii Rashkovskii
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ */
+package graphql.annotations;
+
+import graphql.schema.DataFetcher;
+import graphql.schema.DataFetchingEnvironment;
+import graphql.schema.DataFetchingEnvironmentImpl;
+
+import java.util.Map;
+
+import static graphql.annotations.ReflectionKit.newInstance;
+
+public class ExtensionDataFetcherWrapper<T> implements DataFetcher<T>{
```

Comment:
hey, can you explain to me the need of an ExtensionDataFetcher? I didnt quite understand the purpose of it

Reply:
Sure - the idea was to be able to use the standard FieldDataFetcher and PropertyDataFetcher on extension objects. When using extensions, the source object is still the extended object (in previous example, instance of A), but the fields and getters are on the extension object (class B). The ExtensionDataFetcher is calling the underlying DataFetcher (Property or Field) with an instance of B as the source object.

MethodDataFetcher is already handling that directly (it's part of graphql-java-annotations, as opposed to the Field and Property DataFetcher). It's checking that `method.getDeclaringClass().isInstance(environment.getSource())`, and otherwise tries to build an instance of B, or call static methods on B.

Note that it works only if B has a constructor taking A as a parameter. Otherwise, FieldDataFetcher and PropertyDataFetcher won't be able to retrieve any value. 

You can look at the example in GraphQLExtensionsTest / TestObjectExtension class.

Of course, if we agree on the way to do it, this will need to be documented :-)